sdk/src/types/block/address/restricted.rs

```diff
@@ -0,0 +1,323 @@
+    const DELEGATION_OUTPUTS: u8 = 0b10000000;
+
+    /// Converts the flag into the byte representation.
+    pub fn as_byte(&self) -> u8 {
+        match self {
+            Self::NativeTokens => Self::NATIVE_TOKENS,
+            Self::Mana => Self::MANA,
+            Self::OutputsWithTimelock => Self::OUTPUTS_WITH_TIMELOCK,
+            Self::OutputsWithExpiration => Self::OUTPUTS_WITH_EXPIRATION,
+            Self::OutputsWithStorageDeposit => Self::OUTPUTS_WITH_STORAGE_DEPOSIT,
+            Self::AccountOutputs => Self::ACCOUNT_OUTPUTS,
+            Self::NftOutputs => Self::NFT_OUTPUTS,
+            Self::DelegationOutputs => Self::DELEGATION_OUTPUTS,
+        }
+    }
+
+    /// Returns the index in [`AddressCapabilities`] to which this flag is applied.
+    pub fn index(&self) -> usize {
+        match self {
+            Self::NativeTokens
+            | Self::Mana
+            | Self::OutputsWithTimelock
+            | Self::OutputsWithExpiration
+            | Self::OutputsWithStorageDeposit
+            | Self::AccountOutputs
+            | Self::NftOutputs
+            | Self::DelegationOutputs => 0,
+        }
+    }
+
+    /// Returns an iterator over all flags.
+    pub fn all() -> impl Iterator<Item = Self> {
+        [
+            Self::NativeTokens,
+            Self::Mana,
+            Self::OutputsWithTimelock,
+            Self::OutputsWithExpiration,
+            Self::OutputsWithStorageDeposit,
+            Self::AccountOutputs,
+            Self::NftOutputs,
+            Self::DelegationOutputs,
+        ]
+        .into_iter()
+    }
+}
+
+/// A list of bitflags that represent the capabilities of an [`Address`].
+/// If an output is created by a transaction with an
+/// [`UnlockCondition`](crate::types::block::output::UnlockCondition) containing a [`RestrictedAddress`], the
+/// transaction is valid only if the specified conditions, corresponding to the [`AddressCapabilityFlag`]s, hold for
+/// that output.
+#[derive(Clone, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash, Deref)]
+#[repr(transparent)]
+pub struct AddressCapabilities(BoxedSlicePrefix<u8, u8>);
+
+impl AddressCapabilities {
+    /// Returns an [`AddressCapabilities`] with every possible flag enabled.
+    pub fn all() -> Self {
+        let mut res = Self::default();
+        res.set_all();
+        res
+    }
+
+    /// Returns an [`AddressCapabilities`] with every possible flag disabled.
+    pub fn none() -> Self {
+        Self::default()
+    }
+
+    /// Returns whether every possible [`AddressCapabilityFlag`] is enabled.
+    pub fn is_all(&self) -> bool {
+        AddressCapabilityFlag::all().all(|flag| self.has_capability(flag))
+    }
+
+    /// Returns whether every possible [`AddressCapabilityFlag`] is disabled.
+    pub fn is_none(&self) -> bool {
+        self.0.iter().all(|b| 0.eq(b))
+    }
+
+    /// Enables every possible [`AddressCapabilityFlag`].
+    pub fn set_all(&mut self) -> &mut Self {
```

Comment:
why also return a `&mut Self`? :thinking:

Reply:
borrow checker is fine 😂 it's just for usability so you can chain the fns if you want.

Reply:
borrow checker is fine?
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6349a48ecfcba2fa27aae191032ba6ba

Reply:
I guess it's not such big of a deal and `std` does it as well, so :shrug: (I am mostly used to see chaining with passing ownership along, but that's not possible here, otherwise you'ld probably done it)